### PR TITLE
Feature: Draconic Sacrifice Tracker

### PR DESCRIPTION
**Pull request by superhize**
## What
Tracker for the Draconic Altar sacrifices

<details><summary>Images</summary>
<p>

Hypacrosaurus_t9pgMtPhjI
Thrasher_F6fd8LxIow

</p>
</details> 

## Changelog New Features
+ Added Draconic Sacrifice Tracker. - HiZe
    * Tracks items and profit while using the Draconic Altar in the End.

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        @Expose
+        var itemsSacrifice = 0
+
+        @Expose
+        var essences = 0
+
+        @Expose
+        var bonusRewards: MutableMap<BonusReward, Long> = mutableMapOf();
```
Comment (superhize) [inline src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt]: i'm tired

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/DraconicSacrificeTrackerConfig.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import com.google.gson.annotations.Expose;
+import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorBoolean;
+import io.github.notenoughupdates.moulconfig.annotations.ConfigLink;
+import io.github.notenoughupdates.moulconfig.annotations.ConfigOption;
+
+public class DraconicSacrificeTrackerConfig {
+
+    @Expose
+    @ConfigOption(name = "Enable", desc = "Enable the Draconic sacrifice tracker.")
+    @ConfigEditorBoolean
```
Comment (jani270) [inline src/main/java/at/hannibal2/skyhanni/config/features/misc/DraconicSacrificeTrackerConfig.java:14]: Missing feature toggle

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/DraconicSacrificeTrackerConfig.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import at.hannibal2.skyhanni.config.core.config.Position;
+import com.google.gson.annotations.Expose;
+import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorBoolean;
+import io.github.notenoughupdates.moulconfig.annotations.ConfigLink;
+import io.github.notenoughupdates.moulconfig.annotations.ConfigOption;
+
+public class DraconicSacrificeTrackerConfig {
+
+    @Expose
+    @ConfigOption(name = "Enable", desc = "Enable the Draconic sacrifice tracker.")
```
Comment (jani270) [inline src/main/java/at/hannibal2/skyhanni/config/features/misc/DraconicSacrificeTrackerConfig.java]:
```suggestion
    @ConfigOption(name = "Enabled", desc = "Enable the Draconic sacrifice tracker.")
```

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java [hunk 4]
```diff
@@ -114,6 +114,9 @@ public class MiscConfig {
     public PatcherCoordsWaypointConfig patcherCoordsWaypoint = new PatcherCoordsWaypointConfig();
 
     @Expose
+    @ConfigOption(name = "Dragonic Sacrifice Tracker", desc = "")
+    @Accordion
+    public DraconicSacrificeTrackerConfig draconicSacrificeTracker = new DraconicSacrificeTrackerConfig();
```
Comment (jani270) [inline src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java]: Seems like a missing empty line here?

File: src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    private val config get() = SkyHanniMod.feature.misc.draconicSacrificeTracker
+    private val patternGroup = RepoPattern.group("misc.draconicsacrifice")
+    private val sacrificeLoot by patternGroup.pattern(
+        "sacrifice",
+        "§c§lSACRIFICE! §r§eYou turned §r(?<item>.*) §r§einto §r§d(?<amount>\\d+) Dragon Essence§r§e!",
+    )
+    private val bonusLoot by patternGroup.pattern(
+        "bonus",
+        "§c§lBONUS LOOT! §r§eYou also received §r(?:§\\w(?<amount>\\d+)?x)?(?: §r)?(?<item>.*) §r§efrom your sacrifice!",
+    )
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt]: please include regex tests

File: src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            return listOf(
+                "§7Dropped §e${timesGained.addSeparators()} §7times.",
+                "§7Your drop rate: §c$dropRate.",
+            )
+        }
+
+        override fun getCoinName(item: TrackedItem) = "§dDragon Essence"
+
+        override fun getCoinDescription(item: TrackedItem): List<String> {
+            val essences = NumberUtil.format(item.totalAmount)
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt]: 'format(Number, Boolean = ...): String' is deprecated. 
Probably use add separators instead

File: src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            return listOf(
+                "§7Sacrificed items give you dragon essence.",
+                "§7You got §6$essences essence §7that way.",
+            )
+        }
+
+        @Expose
+        var itemsSacrifice = 0L
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt]: should be itemsSacrificed

File: src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+        @Expose
+        var sacrifiedItemsMap: MutableMap<String, Long> = mutableMapOf()
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt]: should be sacrificedItemsMap

**Review by CalMWolfs — CHANGES_REQUESTED** [review]

File: src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    private fun drawDisplay(data: Data): List<Searchable> = buildList {
+        addSearchString("§5§lDraconic Sacrifice Profit Tracker")
+        val profit = tracker.drawItems(data, { true }, this)
+
+        add(
+            Renderable.hoverTips(
+                "§b${data.itemsSacrificed.addSeparators()} §6Items Sacrified",
+                data.sacrificedItemsMap.map { (key, value) -> "$key: §b$value" },
```
Comment (ItsEmpa) [inline src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt]: if you already are giving names to key and value, probably better to just name them "item" and "amount"
```suggestion
                data.sacrificedItemsMap.map { (item, amount) -> "$item: §b$amount" },
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+        tracker.addPriceFromButton(this)
+    }
+
+    @SubscribeEvent
+    fun onChat(event: LorenzChatEvent) {
+        sacrificeLoot.matchMatcher(event.message) {
+            val amount = group("amount").toInt()
+            val item = group("item")
+            tracker.addItem("ESSENCE_DRAGON".asInternalName(), amount, command = false)
```
Comment (ItsEmpa) [inline src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt]: initialize essence_dragon as internal name in the class and not inside the function, so you dont have to initialize this every time

File: src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        var sacrificedItemsMap: MutableMap<String, Long> = mutableMapOf()
+    }
+
+    private fun drawDisplay(data: Data): List<Searchable> = buildList {
+        addSearchString("§5§lDraconic Sacrifice Profit Tracker")
+        val profit = tracker.drawItems(data, { true }, this)
+
+        add(
+            Renderable.hoverTips(
+                "§b${data.itemsSacrificed.addSeparators()} §6Items Sacrified",
```
Comment (ItsEmpa) [inline src/main/java/at/hannibal2/skyhanni/features/misc/DraconicSacrificeTracker.kt]:
```suggestion
                "§b${data.itemsSacrificed.addSeparators()} §6Items Sacrificed",
```

**Review by ItsEmpa — CHANGES_REQUESTED** [review]

File: src/main/java/at/hannibal2/skyhanni/config/commands/Commands.kt [hunk 12]
```diff
@@ -166,7 +167,6 @@ object Commands {
         }
     }
 
-    @Suppress("LongMethod")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/commands/Commands.kt]: not sure why this is getting removed, maybe a mistake when merging?

File: src/main/java/at/hannibal2/skyhanni/config/commands/Commands.kt [hunk 13]
```diff
@@ -397,6 +397,12 @@ object Commands {
             callback { ExcavatorProfitTracker.resetCommand() }
         }
 
+        event.register("shresetdraconicsacrificetracker") {
+            description = "Resets the Draconic Sacrifice Tracker."
+            category = CommandCategory.USERS_RESET
+            callback { DraconicSacrificeTracker.resetCommand() }
+        }
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/commands/Commands.kt]: Now that command registration is event based you can just add this to DraconicSacrificeTracker instead

    @HandleEvent
    fun onCommandRegistration(event: CommandRegistrationEvent) {
        event.register("shresetdraconicsacrificetracker") {
            description = "Resets the Draconic Sacrifice Tracker."
            category = CommandCategory.USERS_RESET
            callback { tracker.resetCommand() }
        }
    }

**Review by CalMWolfs — APPROVED** [review]
worked for me

**Review by ItsEmpa — APPROVED** [review]

**Comment by hannibal002** [comment]
updated the config a bit
image

**Pull request merged**